// motion/lib/index.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _assert = require('assert');

var _assert2 = _interopRequireDefault(_assert);

var _sbEventKit = require('sb-event-kit');

var _fs = require('./fs');

var FS = _interopRequireWildcard(_fs);

var _cli = require('./cli');

var _cli2 = _interopRequireDefault(_cli);

var _config = require('./config');

var _config2 = _interopRequireDefault(_config);

var _error = require('./error');

var _helpers = require('./helpers');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { return step("next", value); }, function (err) { return step("throw", err); }); } } return step("next"); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Motion = function () {
  function Motion(projectPath, config) {
    var _this = this;

    _classCallCheck(this, Motion);

    (0, _assert2.default)(config instanceof _config2.default, 'Use Motion.create instead of constructor');

    this.cli = new _cli2.default(config);
    this.config = config;
    this.watching = false;
    this.projectPath = projectPath;
    this.subscriptions = new _sbEventKit.CompositeDisposable();

    this.subscriptions.add(this.cli);
    this.cli.onShouldBuild(_asyncToGenerator(function* () {
      yield _this.build(false);
    }));
    this.cli.onShouldDispose(function () {
      _this.dispose();
    });
  }

  _createClass(Motion, [{
    key: 'exists',
    value: function () {
      var _ref2 = _asyncToGenerator(function* () {
        return yield FS.exists(_path2.default.join(this.projectPath, '.motion.json'));
      });

      function exists() {
        return _ref2.apply(this, arguments);
      }

      return exists;
    }()
  }, {
    key: 'watch',
    value: function () {
      var _ref3 = _asyncToGenerator(function* () {
        var _this2 = this;

        var terminal = arguments.length <= 0 || arguments[0] === undefined ? false : arguments[0];

        if (!(yield this.exists())) {
          throw new _error.MotionError(_error.ERROR_CODE.NOT_MOTION_APP);
        }

        if (terminal) {
          this.cli.activate();
        }

        var _ref4 = yield (0, _helpers.getPundleInstance)(this.cli, terminal, this.projectPath, true, this.config.config, function (error) {
          _this2.cli.log(error);
        });

        var subscription = _ref4.subscription;

        var disposable = new _sbEventKit.Disposable(function () {
          _this2.subscriptions.delete(disposable);
          subscription.dispose();
        });

        this.subscriptions.add(disposable);
        return disposable;
      });

      function watch(_x) {
        return _ref3.apply(this, arguments);
      }

      return watch;
    }()
  }, {
    key: 'build',
    value: function () {
      var _ref5 = _asyncToGenerator(function* () {
        var terminal = arguments.length <= 0 || arguments[0] === undefined ? false : arguments[0];

        if (!(yield this.exists())) {
          throw new _error.MotionError(_error.ERROR_CODE.NOT_MOTION_APP);
        }
        var error = void 0;

        var _ref6 = yield (0, _helpers.getPundleInstance)(this.cli, terminal, this.projectPath, false, this.config.config, function (givenError) {
          error = givenError;
        });

        var subscription = _ref6.subscription;
        var pundle = _ref6.pundle;

        try {
          if (error) {
            throw error;
          }
          var generated = yield pundle.generate(null, {
            sourceMap: false
          });
          yield FS.mkdir(_path2.default.join(this.config.getPublicDirectory(), '_'));
          yield FS.writeFile(_path2.default.join(this.config.getPublicDirectory(), '_/bundle.js'), generated.contents);
        } finally {
          subscription.dispose();
        }
      });

      function build(_x3) {
        return _ref5.apply(this, arguments);
      }

      return build;
    }()
  }, {
    key: 'init',
    value: function () {
      var _ref7 = _asyncToGenerator(function* () {
        if (yield this.exists()) {
          throw new _error.MotionError(_error.ERROR_CODE.ALREADY_MOTION_APP);
        }
        yield FS.mkdir(this.config.getBundleDirectory());
        yield FS.mkdir(this.config.getPublicDirectory());
        yield FS.copy(_path2.default.normalize(_path2.default.join(__dirname, '..', 'template', 'bundle')), this.config.getBundleDirectory());
        yield FS.copy(_path2.default.normalize(_path2.default.join(__dirname, '..', 'template', 'public')), this.config.getPublicDirectory());
        yield this.config.write();
      });

      function init() {
        return _ref7.apply(this, arguments);
      }

      return init;
    }()
  }, {
    key: 'dispose',
    value: function dispose() {
      this.subscriptions.dispose();
    }
  }], [{
    key: 'create',
    value: function () {
      var _ref8 = _asyncToGenerator(function* (projectRoot) {
        return new Motion(projectRoot, (yield _config2.default.create(projectRoot)));
      });

      function create(_x5) {
        return _ref8.apply(this, arguments);
      }

      return create;
    }()
  }]);

  return Motion;
}();

module.exports = Motion;
