// pundle-api/lib/issues.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MessageIssue = exports.FileIssue = undefined;

var _assert = require('assert');

var _assert2 = _interopRequireDefault(_assert);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var VALID_SEVERITIES = new Set(['info', 'warning', 'error']);

var FileIssue = exports.FileIssue = function FileIssue(contents, line, column, message) {
  var severity = arguments.length <= 4 || arguments[4] === undefined ? 'error' : arguments[4];

  _classCallCheck(this, FileIssue);

  this.contents = contents;
  this.line = line;
  this.column = column;
  this.message = message;
  this.severity = severity.toLowerCase();

  (0, _assert2.default)(typeof this.contents === 'string' && this.contents, 'Contents must be a valid string');
  (0, _assert2.default)(typeof this.line === 'number' && this.line, 'Line must be a valid number');
  (0, _assert2.default)(typeof this.column === 'number' && this.column, 'Column must be a valid number');
  (0, _assert2.default)(typeof this.message === 'string' && this.message, 'Message must be a valid string');
  (0, _assert2.default)(VALID_SEVERITIES.has(this.severity), 'Severity must be valid');
};

var MessageIssue = exports.MessageIssue = function MessageIssue(message) {
  var severity = arguments.length <= 1 || arguments[1] === undefined ? 'error' : arguments[1];

  _classCallCheck(this, MessageIssue);

  this.message = message;
  this.severity = severity;
  (0, _assert2.default)(typeof this.message === 'string' && this.message, 'Message must be a valid string');
  (0, _assert2.default)(VALID_SEVERITIES.has(this.severity), 'Severity must be valid');
};
